### PR TITLE
fix(learn): dedup replayed tool calls by tool_call_id in loop detection

### LINKED ISSUE #3452: [BUG] learn loop detection double-counts replayed transcripts (no tool_call_id dedup)
## Description

`headroom learn`'s loop detection counts the same tool call once per transcript file it appears in. `detect_loops()` ([`loops.py:121-130`](https://github.com/headroomlabs-ai/headroom/blob/e59cf1012a2b277689555833fd221a92279f777d/headroom/learn/loops.py#L121-L130)) merges each session's qualifying groups into a global view keyed by signature, and never dedups by `tool_call_id`:

```python
for sig, calls in per_session.items():
    if len(calls) >= min_occurrences:
        groups.setdefault(sig, []).extend(calls)   # <- no id dedup
```

`ToolCall.tool_call_id` carries the provider-assigned id and is already parsed by the scanners, so the information needed to tell a replay from a repetition is present and unused.

When a provider resumes a conversation by writing a **new** transcript that replays prior history, every replayed call is scanned again and counted again. Three reads become six; the measured waste rises with them. Nothing about the agent's behavior changed — only the number of files on disk describing it.

This is not the truncation defect in #3440. That one merges *distinct* calls because the grouping key is lossy; this one over-counts calls that are correctly grouped and genuinely identical, because the same call is presented to the scanner more than once. The two compound: an already-inflated loop is then multiplied by the number of replays.

### Consequence: `min_occurrences` is evaluated before dedup

The threshold check (`len(calls) >= min_occurrences`) runs on the per-session group *before* any deduplication. So a group whose repetitions are entirely replays of a single call clears the bar on replay count alone, and is reported as a loop with a real signature and attributed waste — despite the underlying session containing one call. Any fix that dedups after the threshold has to re-check it, or it converts double-counted loops into phantom ones rather than removing them.

### Consequence: `tool_call_id` uniqueness becomes a scanner contract

Deduping by id means an id must be unique *across* sessions, not merely within one. Where a scanner synthesizes ids rather than reading them from the transcript, two unrelated sessions running the same loop would otherwise look like one session replayed twice — the same defect in the opposite direction. This is worth stating explicitly wherever the dedup lands, since it constrains every current and future plugin.

## To Reproduce

1. `pip install headroom-ai` (originally reproduced on 0.36.5; the code path is unchanged on `main`)
2. Save the script below as `repro_dedup.py`
3. `python repro_dedup.py`

## Code Sample

```python
"""Repro: headroom learn loop detection never dedups by tool_call_id."""
import headroom
from headroom.learn.loops import detect_loops
from headroom.learn.models import SessionData, ToolCall

print("headroom", headroom.__version__)

def tc(name, tool_call_id, input_data, out_bytes=40000, msg_index=0):
    return ToolCall(name=name, tool_call_id=tool_call_id, input_data=input_data,
                    output="x" * out_bytes, is_error=False,
                    msg_index=msg_index, output_bytes=out_bytes)

# The same three calls, seen once, then seen again in a resume transcript that
# replays them. Identical tool_call_ids in both -- the same calls, not new ones.
reads = [tc("Read", f"call_r{i}", {"file_path": "/repo/docs/design.md"}, msg_index=i)
         for i in range(3)]
one = detect_loops([SessionData(session_id="rollout-1", tool_calls=list(reads))])
fork = detect_loops([SessionData(session_id="rollout-1", tool_calls=list(reads)),
                     SessionData(session_id="rollout-2-resume", tool_calls=list(reads))])
print("\n[2] same 3 calls, single transcript vs replayed resume transcript:")
print(f"    single : count={one[0].count} wasted={one[0].wasted_tokens:,}")
print(f"    resumed: count={fork[0].count} wasted={fork[0].wasted_tokens:,}")
```

## Expected Behavior

| Input | Expected |
|---|---|
| 3 calls, replayed verbatim in a resume transcript | `count=3`, `wasted=20,000` — same as the single-transcript case |
| 3 calls, plus 2 genuinely new calls in the resume | `count=5` — a resume that continues the loop still accumulates |
| 3 calls carrying no `tool_call_id`, replayed | `count=6` — nothing identifies them as replays, so they stay counted; this case is deliberately not fixed |

## Actual Behavior

```
headroom 0.36.5

[2] same 3 calls, single transcript vs replayed resume transcript:
    single : count=3 wasted=20,000
    resumed: count=6 wasted=50,000
```

The replay adds three phantom repetitions and 30,000 phantom wasted tokens. Inflated counts propagate into `format_loops_for_digest()` (the "HIGHEST PRIORITY" LLM section) and `apply_loop_weighting()`, so a loop that is merely replayed can outrank one that is real.

## Environment

- **Headroom version**: 0.36.5 (pipx); re-verified against `main` @ `e67b3c8a` (reports `0.38.0-dev`), where the script above prints the identical `single: count=3 / resumed: count=6`.
- **Python version**: 3.13
- **OS**: Ubuntu (WSL2)
- **LLM Provider**: Anthropic (Claude Code CLI)

## Proposed fix — PR incoming

- Dedup by `tool_call_id` when merging per-session groups into the global view. Calls carrying **no** id stay counted, since nothing identifies them as replays.
- Re-check `min_occurrences` after dedup, so a group that only cleared the bar via replays is dropped rather than reported as a sub-threshold loop.
- Write the cross-session uniqueness requirement into `detect_loops()`'s docstring, and scope synthesized ids by session in the scanners that generate them.

Behavior-preserving for the builtin path pinned by `tests/test_learn/test_loop_weighting.py`. I'm opening a PR with these plus regression tests (failing before, passing after) and the real-behavior proof the contributing guide asks for.

## Related

- #3440 — the grouping-key truncation defect this was split out of. Different mechanism, and the two compound; fixed in a separate PR.
- #1159 — the feature this module came from (loop weighting). This is an accuracy follow-up, not a regression report against it.

### PR DESCRIPTION
**Pull request by aniruddhaadak80**
## Description

`detect_loops()` merged each session's qualifying signature groups into the global view without deduplicating by `tool_call_id`, so a provider resume transcript that replays prior history double-counted every replayed call (3 reads became 6, waste 20,000 became 50,000).

Root cause: `headroom/learn/loops.py` grouped by canonical signature per session and extended the global `groups` dict directly, never consulting `ToolCall.tool_call_id` even though scanners already parse it.

Fixes #3452

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring (no functional changes)

## Changes Made

- `headroom/learn/loops.py:121-129` (docstring): documented the `tool_call_id` dedup contract, the empty-id exception, the post-dedup `min_occurrences` re-check, and the cross-session uniqueness requirement for synthesized scanner ids.
- `headroom/learn/loops.py:132` (`seen_ids: set[str]`): global dedup set across sessions.
- `headroom/learn/loops.py:141-152`: when merging per-session qualifying groups, skip calls whose stripped non-empty `tool_call_id` was already seen; calls with empty ids are never deduped.
- `headroom/learn/loops.py:154-156`: re-check `min_occurrences` after dedup so a group that only cleared the bar via replays is dropped instead of reported as a phantom sub-threshold loop.
- `tests/test_learn/test_loop_weighting.py:92-152` (`_read_call` helper + `TestDetectLoopsDedup` with 4 tests): replayed-resume stays at count 3, resume with 2 genuinely new calls accumulates to 5, empty-id replays stay counted at 6 (deliberately unfixed case), pure-replay single-id triple is dropped below threshold.

## Testing

- [x] Unit tests pass (`pytest`)
- [x] Linting passes (`ruff check .`)
- [x] Formatting passes (`ruff format --check`)
- [ ] Type checking passes (`mypy headroom`)
- [x] New tests added for new functionality
- [x] Manual testing performed

### Test Output

```text
# Target suite after fix (17/17 green):
tests\test_learn\test_loop_weighting.py .................  [100%]
17 passed, 1 warning in 6.10s

# New regression tests against PRE-FIX code (only headroom/learn/loops.py stashed):
tests\test_learn\test_loop_weighting.py FF.F  [100%]
FAILED test_replayed_resume_transcript_not_double_counted - assert 6 == 3
  (LoopPattern count=6, wasted_tokens=50000, msg_indices=[0, 0, 1, 1, 2, 2])
FAILED test_resume_with_genuinely_new_calls_accumulates - assert 8 == 5
FAILED test_pure_replay_duplicates_dropped_below_threshold - assert [LoopPattern(count=3 ...)] == []
3 failed, 1 passed (test_calls_without_ids_stay_counted passes before and after by design)

# Full tests/test_learn/ after fix:
6 failed, 245 passed, 4 skipped in 33.37s
Failures (all in files NOT touched by this PR):
- tests/test_learn/test_analyzer.py::TestDigestBuilder::test_includes_project_info
- tests/test_learn/test_integration.py::TestIdempotency::test_double_write_replaces_not_appends
- tests/test_learn/test_scanner.py (4x TestDecodeProjectPath Windows-path greedy decoding)
Honest scoping: a clean-tree re-run of those 6 to prove pre-existing status timed out
past 120s in this environment, so pre-existing is inferred from file isolation
(none touch loops.py or loop_weighting.py), not proven here. Leaving to CI.

# Lint / format (after fix):
uvx ruff check headroom/learn/loops.py tests/test_learn/test_loop_weighting.py
All checks passed!
uvx ruff format --check headroom/learn/loops.py tests/test_learn/test_loop_weighting.py
2 files already formatted
```

Note: `mypy` was not run (dev extra not installed in this Windows environment).

## Real Behavior Proof

- Environment: Windows 11, Python 3.13.13, uv 0.11.14, repo at `upstream/main` (`e67b3c8`) + this patch.
- Exact command / steps:
  1. Saved the issue's `repro_dedup.py` verbatim to `%TEMP%\opencode\repro_dedup.py`.
  2. `$env:PYTHONPATH='<repo>'; uv run --no-sync python -u <repro_dedup.py>`
- Observed result (after fix):
```text
headroom 0.38.0-dev
[2] same 3 calls, single transcript vs replayed resume transcript:
    single : count=3 wasted=20,000
    resumed: count=3 wasted=20,000
```
Matches the issue's Expected Behavior row 1 (`count=3`, `wasted=20,000`). The issue reports pre-fix output as `single: count=3 / resumed: count=6, wasted=50,000`; the stashed-code pytest failures above (`assert 6 == 3`, `wasted 50000`) confirm the same pre-fix shape on this checkout.
- Not tested: end-to-end `headroom learn` CLI over real transcript files; scanners that synthesize ids (contract documented in the docstring but no scanner code changed here).

## Runtime Rollout Safety

- Rollout-managed feature(s): none
- Minimum rollout channel: n/a
- Stable/default behavior changed: yes, intentionally — replayed `tool_call_id`s no longer inflate loop counts/waste. Single-session loops with unique ids and all empty-id paths are unchanged (pinned by `test_calls_without_ids_stay_counted` and the 13 pre-existing tests).
- Kill switch / disable path: n/a
- Unsafe override required: no
- Qualification impact: `format_loops_for_digest()` and `apply_loop_weighting()` now receive deduplicated counts, so replayed loops no longer outrank real ones.
- Rollback path: revert this commit

## Review Readiness

- [x] I have performed a self-review
- [x] This PR is ready for human review

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (docstring updated; no wiki/docs change needed)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes (target suite 17/17; full learn suite has 6 unrelated failures noted honestly above)
- [x] I did **not** edit `CHANGELOG.md` — it is generated by release-please from my Conventional Commit PR title (a CI guard enforces this)

## Screenshots (if applicable)

n/a

## Additional Notes

- Scanner contract (from #3452): dedup by id requires ids unique *across* sessions. Where a scanner synthesizes ids, it must scope them by session id, otherwise two unrelated sessions running the same loop would collapse into one — the same defect in reverse. Stated in the `detect_loops()` docstring; no scanner code changed in this PR.
- One logical change per PR; diff is `headroom/learn/loops.py` (+28/-1) and `tests/test_learn/test_loop_weighting.py` (+62). Not requesting reviews from specific humans.

### CODE REVIEW

File: headroom/learn/loops.py [hunk 1]
```diff
@@ -127,7 +138,22 @@ def detect_loops(
...
+                    if tid in seen_ids:
+                        continue
+                    seen_ids.add(tid)
+                    unique.append(tc)
+                if unique:
+                    groups.setdefault(sig, []).extend(unique)
+
+    # Re-check the threshold after dedup: a group that only cleared the bar
+    # via replayed calls must not surface as a sub-threshold phantom loop.
+    groups = {sig: calls for sig, calls in groups.items() if len(calls) >= min_occurrences}
```
Comment (JerrettDavis) [inline headroom/learn/loops.py:156]: [P2] Re-check each session after local deduplication, before combining sessions. With min_occurrences=3, session A has same-signature call IDs [a1,a1,a2] and unrelated session B has [b1,b1,b2]. Each session correctly returns no loops by itself, but detect_loops([A,B]) now returns count=4 and 30,000 wasted tokens for 40KB outputs. Neither session had three real calls; the final global threshold lets their two sub-threshold groups qualify together. Please first deduplicate within each session and apply its threshold, then apply cross-transcript replay deduplication when accumulating qualifying groups. Add this two-session regression alongside the existing resume-with-new-calls test.

**Review by JerrettDavis — CHANGES_REQUESTED** [review]
The replay-only and resumed-history regressions pass (17 focused tests locally), and deduplication is a useful correction. One threshold edge case still produces a phantom loop across unrelated sessions; please keep the within-session qualification invariant after removing local duplicates.

**Comment by JerrettDavis** [comment]
Thanks for digging into replayed tool calls. I’m consolidating the fix for #3452 in #3462, which also handles the cross-session threshold case raised in review and keeps synthetic call IDs distinct across sessions. Closing this alternative so that correction has one place to land.
